### PR TITLE
accounts: Don't collect rent on newly created accounts

### PR DESCRIPTION
**Pull request by joncinque**
#### Problem

This is a doozy for a one-line change.

With  #26479 activated, the rent epoch of accounts is no longer updated. If you start a new cluster (or test validator) with this enabled, there may be some accounts with `rent_epoch = 0`.

When one of those accounts is closed, the runtime thinks it's a new account since its rent epoch is 0, tries to collect rent on it, and cleans up the account in the process.

This is normally OK, but if the account in question is a stake account, the stakes cache will go out of sync and eventually crash the bank during epoch rollover on this assertion: https://github.com/solana-labs/solana/blob/422cff69fdc6e9268020dd52db842c6a0b64a4cd/runtime/src/bank.rs#L2704

#### Summary of Changes

Very simply, don't collect rent on newly created accounts if the feature flag from #26479 is enabled.  Since newly created accounts cannot be rent-paying, this rent collection never happened anyway.  The rent collector code can be cleaned up once the feature is removed.

Fixes #

### CODE REVIEW

**Comment by jeffwashington** [comment]
https://github.com/solana-labs/solana/blob/9ef684f5364a5772f0e9ef92d873e141c4d2332e/runtime/src/accounts.rs#L1303-L1307

I don't understand the ramifications. Is it this line that is the problem? rent should be 0.

**Comment by joncinque** [comment]
@jeffwashington The issue is actually here https://github.com/solana-labs/solana/blob/9ef684f5364a5772f0e9ef92d873e141c4d2332e/runtime/src/rent_collector.rs#L142 -- we collect rent on any newly created accounts, and delete any of them that don't have enough lamports for rent.  Since accounts created at epoch 0 never have their rent epoch advanced, when closing these accounts, they'll get re-initialized too early.

**Comment by jeffwashington** [comment]
> when closing these accounts, they'll get re-initialized too early.

Closing means setting lamports to 0?
How would we be re-initializing them (setting AccountSharedData to default?) 'too' early?

**Comment by joncinque** [comment]
> Closing means setting lamports to 0?

Correct

> How would we be re-initializing them (setting AccountSharedData to default?) 'too' early?

The account is re-initialized before updating the stakes cache. If it's a stake account, we lose any delegation information, causing the stakes cache to go out of sync.

File: runtime/src/accounts.rs
Comment (behzadnouri) [inline runtime/src/accounts.rs:1293]: When do we still need to take this branch?

We no longer create rent-paying accounts; so in what scenario do we still need to take this branch and collect rent on a new account?

Comment (joncinque) [reply runtime/src/accounts.rs:1293]: This branch is still taken when an account with rent epoch 0 is referenced in a transaction.  On every transaction, there's a rent collection that takes place for new accounts, and we check if an account is new based on `rent_epoch == 0`.

However, if an account has been closed (emptied of lamports), we still try to do this rent collection afterwards, and since that account now owes rent, the runtime reinitializes the account to `Account:;default()`.  This is mostly OK, but for stake accounts, this means that the cache falls out of sync, since the cache update happens after the reinitialization.

Comment (joncinque) [reply runtime/src/accounts.rs:1293]: No reason that I can see, it was simply just never removed.  Do you prefer just removing this whole branch rather than tying it to your feature?

Comment (behzadnouri) [reply runtime/src/accounts.rs:1293]: yeah, _if that is the case than the branch is never taken anyways_, then lets remove it entirely.
cc @CriesofCarrots to confirm that the branch is redundant.

Comment (CriesofCarrots) [reply runtime/src/accounts.rs:1293]: Checking! It might be safest to feature-gate the removal, but depending on the order of operations here, we can maybe convince ourselves that's not necessary

Comment (CriesofCarrots) [reply runtime/src/accounts.rs:1293]: Okay, I have confirmed that when we reach this line, all transactions that leave any account in a newly rent-paying state will:
a. have `execution_status.is_err()` (`TransactionError::InvalidRentPayingAccount` or `TransactionError::InsufficientFundsForRent`)
b. not feature a rent-paying nonce account, because SystemProgram enforces this
c. not feature a rent-paying fee_payer account, because such a transaction will have result `TransactionExecutionResult::NotExecuted` and not reach this line

Thus any tx with a rent-paying account will not follow this branch.

Transactions that create rent-exempt accounts still take the branch, and call `collect_from_created_account` which includes https://github.com/solana-labs/solana/blob/e74ad90cdb574c8b8598418a4591804e8cf61b33/runtime/src/rent_collector.rs#L203
So I think the feature gating in this PR is still necessary in order to ensure consistency of the rent_epoch field on account creation when nodes update to the change set.

Handling of RentPaying cases (ie calling `collect_from_existing_account`) should not be necessary at all, though.

Comment (joncinque) [reply runtime/src/accounts.rs:1293]: Thanks so much for looking into all this! And great catch on setting the rent epoch.  @behzadnouri are you ok with using your existing feature flag? Since we're no longer advancing rent epoch with your change, it makes sense to also not set rent epoch at creation with the same feature, but I can be convinced otherwise.

Comment (jeffwashington) [reply runtime/src/accounts.rs:1293]: > Since we're no longer advancing rent epoch with your change, it makes sense to also not set rent epoch at creation with the same feature, but I can be convinced otherwise.

This sounds correct to me.

Comment (CriesofCarrots) [reply runtime/src/accounts.rs:1293]: Then we can remove the block entirely on feature cleanup :)

Comment (behzadnouri) [reply runtime/src/accounts.rs:1293]: @joncinque yeah, sounds good.

**Review by behzadnouri — APPROVED** [review]

**Comment by joncinque** [comment]
@jeffwashington do this get your explicit approval too? I want to be sure before merging.

**Comment by jeffwashington** [comment]
Yes

**Comment by joncinque** [comment]
Thank you for all the gracious reviews!

**Pull request merged**
